rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+  - `AuthorizationHook`: Called when a router request is authorized.
+- TelemetryHooks: Provides hooks for OpenTelemetry tracing and metrics.
+  - `TelemetrySpanHook`: Called when a span is created.
+  - `TelemetryMetricHook`: Called when a metric is recorded.
+- GraphServerHooks: Provides hooks for the lifecycle of a GraphQL server.
+  - `GraphServerStartHook`: Called when the GraphQL server starts e.g. for the first time or when the schema is updated.
+  - `GraphServerStopHook`: Called when the GraphQL server stops e.g. when the application is shut down or the old schema is replaced.
+- GraphQLOperationHooks: Provides hooks for parsed, normalized, and planned GraphQL operations.
+  - `GraphQLOperationParseHook`: Called when an operation is parsed.
+  - `GraphQLOperationNormalizeHook`: Called when an operation is normalized.
+  - `GraphQLOperationPlanHook`: Called when an operation is planned.
+- `ModuleHooks` (**Required**): Provides hooks for the module lifecycle, including provisioning and shutdown.
+  - `Provision`: Called when the module is provisioned.
+  - `Shutdown`: Called when the module is shutdown.
+  - `Module`: Returns the module information and factory function.
+
+For some hooks, we will also provide `pre/post` hooks. This allows developers to perform additional logic after the main hook has been executed e.g. to annotate a span with additional attributes or events.
+
+```go
+type MyModule struct{}
+
+type Subgraph struct {
+	// The name of the subgraph
+	Name string
+	// The ID of the subgraph
+	ID string
+	// The URL of the subgraph
+	URL *url.URL
+	// The schema of the subgraph
+	Schema *graphql.Schema
+}
+
+type Federatedgraph struct {
+	// The name of the graph
+	Name string
+	// The ID of the graph
+	ID string
+	// Version of the graph
+	Version string
+	// Subgraphs of the graph
+	Subgraphs []*core.Subgraph
+	// GraphQL schema of the graph
+	Schema *graphql.Schema
+}
+
+type Operation struct {
+	// The name of the operation
+	Name string
+	// The type of the operation (query, mutation, subscription)
+	Type string
+	// The document representation of the operation
+	Document graphql.Document
+	// The variables of the operation
+	Variables map[string]interface{}
+	// The client name of the operation
+	ClientName string
+	// The client version of the operation
+	ClientVersion string
+	// The uploaded files of the operation
+	UploadedFiles map[string]*core.UploadedFile
+	// The plan of the operation
+	Plan *core.Plan
+	// The normalization of the operation
+	Normalization *core.Normalization
+}
+
+type RouterRequest struct {
+	// The original GraphQL request with all the information like query, variables, operation name, extensions etc.
+	Request *core.GraphQLRequest
+	// The parsed, normaliazed and planned operation with all the information like name, variables, type, document representation,
+	// client name version, uploaded files, plan, normalization, persisted operation etc.
+	Operation *core.Operation
```

Comment:
Somewhere in here I wanted to be able to get
```type Subgraph struct {
	Id string
	Name string
	Url *url.URL
	UrlString string
}
```

So that this information can be used when changing the error.